framework/src/vectorpostprocessors/CSVReader.C

```diff
@@ -65,10 +66,16 @@ CSVReader::CSVReader(const InputParameters & params)
 void
 CSVReader::initialize()
 {
-  _csv_reader.read();
-  for (auto & name : _csv_reader.getNames())
-    if (_column_data.find(name) == _column_data.end())
+  if (_column_data.empty())
+  {
+    _csv_reader.read();
+    for (auto & name : _csv_reader.getNames())
       _column_data[name] = &declareVector(name);
+  }
+  else if (!getParam<bool>("force_preic") ||
+           (getParam<bool>("force_preic") &&
+            (_fe_problem.getCurrentExecuteOnFlag() != EXEC_INITIAL)))
+    mooseError("Error in " + name() + ". CSVReader cannot execute more than once per file.");
```

Comment:
This works today but if we mess with flags a bit more this could break, because it breaks encapsulation principles. 

A better solution, would be to just have a member in this class (or the base) called "_has_executed" or something that gets set to true when it executes and then you check it to see if it's already been set.

Reply:
Agree. 

@crswong888 would you make these changes? Please reference this PR number. If you can't I will do it, but it will take me a bit of time to get to it. Thanks.

Reply:
@aeslaughter  Yes, I can make these changes. My basic understanding of the `contains_complete_history` parameter was that it related to appending more data to the vectors over time, and since were only executing once, why even have this parameter. Although I’m not sure if true or false would be the correct default. Also I’ll switch to the _has_executed check instead.

Reply:
The "contains_complete_history = true" will also disable creating an output file for each timestep. If this object is output, then only a single file is created.

It might be worth setting in "outputs = 'none'" as well in validParams. This way it will not output unless the user enables it. I can't see the purpose of wanting to output a file that you read in.

Thanks

Reply:
I agree, except for the purpose of CSVDiff tests. Is there a way to make an exception for those?

Reply:
or perhaps the data from CSVReader could be passed to another object to output the comparison data on those tests...